### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by joel-kalema**
#### Integration specs
-  Use Capybara to write integration tests for each view in your project.
  
- User index page:
  - I can see the username of all other users.
  - I can see the profile picture for each user.
  - I can see the number of posts each user has written.
  - When I click on a user, I am redirected to that user's show page.
  
- user show page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see the user's bio.
  - I can see the user's first 3 posts.
  - I can see a button that lets me view all of a user's posts.
  - When I click a user's post, it redirects me to that post's show page.
  - When I click to see all posts, it redirects me to the user's post's index page.

- User post index page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see a post's title.
  - I can see some of the post's body.
  - I can see the first comments on a post.
  - I can see how many comments a post has.
  - I can see how many likes a post has.
  - I can see a section for pagination if there are more posts than fit on the view.
  - When I click on a post, it redirects me to that post's show page.

- Post show page:
  - I can see the post's title.
  - I can see who wrote the post.
  - I can see how many comments it has.
  - I can see how many likes it has.
  - I can see the post body.
  - I can see the username of each commentor.
  - I can see the comment each commentor left.

### CODE REVIEW

File: spec/features/users_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    # add five posts for the user
+    5.times do |i|
+      Post.create(author: @user, title: "Post #{i}", text: 'This is my new post', comments_counter: 0, likes_counter: 0)
+    end
+    visit user_path(@user.id)
+  end
+
+  it 'displays users profile picture' do
+    find('img[alt="Jake"]')
+  end
```
Comment (AnselemOdims) [inline spec/features/users_show_spec.rb:17]: - [x] Please kindly assert using `expect` to confirm that the found profile picture is on the page

File: spec/features/users_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    expect(page).to have_content(@user.name)
+  end
+
+  it 'show number of posts written by the user' do
+    expect(page).to have_content("Number of posts: #{@user.posts_counter}")
+  end
+
+  it 'the users bio is visible' do
+    expect(page).to have_content(@user.bio)
+  end
```
Comment (AnselemOdims) [inline spec/features/users_show_spec.rb:33]: - [x] Please kindly write test to assert that you can see the user's first 3 posts.

File: spec/features/users_show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+  it 'button to view all posts is visible' do
+    find_button('See all posts')
+  end
```
Comment (AnselemOdims) [inline spec/features/users_show_spec.rb]: - [x] Please kindly assert that this button can be found on the page

File: spec/features/users_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+  it 'the users bio is visible' do
+    expect(page).to have_content(@user.bio)
+  end
+end
```
Comment (AnselemOdims) [inline spec/features/users_show_spec.rb:55]: - [x] Please kindly include tests for the following assertions
 - It redirects to the post's show page when you click on a user's post
 - It redirects to the user's post's index page when you click on a **see all posts**.

File: spec/features/posts_index_spec.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  end
+
+  it 'display the post body' do
+    expect(page).to have_content(@first_post.text)
+  end
+
+  it 'display the first comment on a post' do
+    expect(page).to have_content('Hi Lilly!')
+  end
+end
```
Comment (AnselemOdims) [inline spec/features/posts_index_spec.rb:52]: - [x] Please kindly include tests for the following assertions
 - asserts you can see the number of posts the user has written
 - asserts you can see how many comments a post has.
 - asserts you can see how many likes a post has.
 - asserts you can see a section for pagination if there are more posts than fit on the view.
 - redirects to the post's show page when you click on a post.

File: spec/features/posts_show_spec.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  end
+
+  it 'I can see the username of each commentor' do
+    expect(page).to have_content(@user2.name)
+  end
+
+  it 'I can see the comment each commentor left' do
+    expect(page).to have_content('Hi Lilly!')
+  end
+end
```
Comment (AnselemOdims) [inline spec/features/posts_show_spec.rb:42]: - [x] Please kindly include tests for the following assertions
 - asserts you can see who wrote the post.
 - asserts you can see how many comments it has.
 - asserts you can see how many likes it has.

**Review by AnselemOdims — CHANGES_REQUESTED** [review]
### Hi @joel-kalema , 
You have been doing a tremendous job so far :ok_hand:
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight 🎉
PR has a title and summary. :heavy_check_mark:
No linter errors :heavy_check_mark:

## Required Changes ♻️
- [ ] Please also check inline comments under the review

## Optional suggestions
_Comments with the [OPTIONAL] prefix are not a must to implement however, It's strongly recommended to work on them as they can make your code better._

## Happy coding!👏👏👏

#### Please leave any questions or comments in the PR if you have any doubts or unclarity! You can tag me using @AnselemOdims for a quicker response

**_Note: Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Review by prolajumokeoni — APPROVED** [review]
Hi @joel-kalema, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
## Highlights
- No linter errors ✅ 
- Git flow was used correctly 🚀 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification @prolajumokeoni.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**